### ISSUE TITLE
wslbridge2 "Exit 1" on non-interactive shell

### ISSUE THREAD

**Issue by iuliantatarascu** [mintty/wsltty#254]
# steps to reproduce:
1. run `C:/Users/IT/AppData/Local/wsltty/bin/mintty.exe '--WSL=' --hold error --exec pwd`

# current behavior:

the output is:
```
/bin/wslbridge2: Exit 1.
/home/it/Desktop
```

wslbridge2 shows an error

window stays open because wslbridge2 returns 1, and because `--hold error` flag is present

# expected behavior:

the output should be:
```
/home/it/Desktop
```

wslbridge2 error should not be present

the window should close because `pwd` command or shell script returns 0 status

the window should stay open only when the `--exec` command returns non-zero status

# additional details:
WSL version 1
Windows 10 version 2004
mintty version 3.3.0
wsltty version 3.3.0

**Comment by mintty** [comment]
If you don't run a shell, your path is not set up as usual. Try `--exec /bin/pwd`.

**Comment by iuliantatarascu** [comment]
the command runs correctly with and without absolute path, but the wslbridge2 error is still present

**Comment by mintty** [comment]
@Biswa96:
This seems to be a wslbridge2 issue; it returns exit code 1 even if the server process succeeds. Maybe it should propaget the backend exit code back like ssh does.

**Comment by Biswa96** [comment]
ACK.

**Comment by dezza** [comment]
I just updated to `Microsoft Windows [Version 10.0.20211.1000]` fastring (it was not an issue with 20206)
```
/bin/wslbridge2: Exit 1.
GetVmId: The parameter is incorrect.
```
@Biswa96 does it look like GetVmId is broken?

**Comment by Biswa96** [comment]
I have not tested the new Windows Insider release. GetVmId depends on undocumented COM interface, so it can be broken.

**Comment by dezza** [comment]
Is GetVmId essential to shell invocation?

**Comment by Biswa96** [comment]
Yes, SOCKADDR_HV needs it. Without it wslbridge2 == wslbridge.

**Comment by Biswa96** [comment]
Yes, SOCKADDR_HV needs it. Without it wslbridge2 == wslbridge. If possible, I will try to update to new Windows insider build tonight.

**Comment by Biswa96** [comment]
Yes, SOCKADDR_HV needs it. Without it wslbridge2 == wslbridge.

**Comment by dezza** [comment]
I reverted to 20206 and it indeed solved the issue. So the problem is verified to be 20211

**Comment by mintty** [comment]
Can the information of GetVmId also be looked up in the registry?
I know, you don't like the registry, but as GetVmId is undocumented, too, and if it's broken, that could be a working fallback.

**Comment by Biswa96** [comment]
> Can the information of GetVmId also be looked up in the registry?

No, the VM GUID is not in registry. It is created randomly by LxssManager service with CoCreateGuid. And the GUID is changed when WSL2 VM restarts.

**Comment by Biswa96** [comment]
> Can the information of GetVmId also be looked up in the registry?

No, the VM GUID is not in registry. It is created randomly by LxssManager service with CoCreateGuid. And the GUID is changed when WSL2 VM restarts. Though there are two possible way to circumvent the COM usage.

1. Add a [integration service][1]. But it requires to create a registry with admin permission.
2. Another way is to use [Hyper-V Compute Service APIs][2]. HcsCreateOperation() -> HcsEnumerateComputeSystems(). But it requires admin permission and a json parser.

[1]: https://docs.microsoft.com/en-us/virtualization/hyper-v-on-windows/user-guide/make-integration-service
[2]: https://docs.microsoft.com/en-us/virtualization/community/team-blog/2017/20170127-introducing-the-host-compute-service-hcs

**Comment by Biswa96** [comment]
Just tested with Windows 10 insider build 20211. I will add the change in wslbridge2 repository. If you can't wait, just add this diff to wslbridge2 master branch.

<details><summary>Click to open the diff</summary>

```diff --git a/src/GetVmId.cpp b/src/GetVmId.cpp
index a1f47a5..272de10 100644
--- a/src/GetVmId.cpp
+++ b/src/GetVmId.cpp
@@ -82,6 +82,7 @@ public:
         /*_In_opt_*/ USHORT WindowHeightY,
         /*_In_*/ ULONG ConsoleHandle,
         /*_In_*/ PLXSS_STD_HANDLES StdHandles,
+        ULONG x,
         /*_Out_*/ GUID *InitiatedDistroId,
         /*_Out_*/ GUID *LxInstanceId,
         /*_Out_*/ PHANDLE LxProcessHandle,
@@ -148,6 +149,7 @@ HRESULT GetVmId(
             nullptr, 0, nullptr, 0, 0,
             HandleToULong(ConsoleHandle),
             &StdHandles,
+            0,
             &InitiatedDistroID,
             LxInstanceID,
             &LxProcessHandle,
```
</details>

**Comment by mintty** [comment]
Hi @Biswa96,
before you make further changes to wslbridge2, please consider that patch I made in fork https://github.com/mintty/wslbridge2 to fix #220. I have configured wsltty to use that version for now, not the one from your repository, to have that patch in.

**Comment by Biswa96** [comment]
I have some doubt about it. Maybe it's my opinion, I like not to process the buffer. If I make a different branch containing that fix, will it be ok? I shall sync both my branches. If I have time I would like to test some ideas. If those failed I shall add your patch.

**Comment by mintty** [comment]
I understand your principal concern so let me elaborate a bit.
I called my approach "escaped inband information" after inband signalling which is an established concept e.g. in telephony networks. I crafted the implementation very carefully not to risk any trouble; it's part of wsltty since 3.1.8 and there has been no complaint. I'm not completely happy with the approach myself but I found no other way after a bunch of experiments.
I you prefer to take it as a branch and keep it synched, that's OK for me. I just wouldn't like to take the burden of maintaining branches or continuous patches on my side.

**Comment by ezillinger** [comment]
@Biswa96 can you share your wslbridge2 build with the fix? I'm not itching to install cygwin.

Thanks!

**Comment by Biswa96** [comment]
You could download the artifacts from [wslbridge2 Appveyor CI][1]. Extract wslbridge2.exe from the 7zip file and place it in `C:\Users\UserName\AppData\Local\wsltty\bin` path, if you have installed wsltty already.

[1]: https://ci.appveyor.com/project/Biswa96/wslbridge2

**Comment by Biswa96** [comment]
You could download the artifacts from [wslbridge2 Appveyor CI][1]. Extract wslbridge2.exe from the 7zip file and place it in `C:\Users\UserName\AppData\Local\wsltty\bin` path, if you have installed wsltty already.

BTW, if anyone is using Windows 10 builds older than 20211, please test it also.

[1]: https://ci.appveyor.com/project/Biswa96/wslbridge2

**Comment by ezillinger** [comment]
@Biswa96 Thank you! Works like a charm in build 20215.

**Comment by mintty** [comment]
@Biswa96, I've just released mintty 3.4.0. Please notify when wslbridge2 is ready to be integrated for wsltty 3.4.0, as an update in the fix-window-resize branch.

**Comment by Biswa96** [comment]
Thank you. I shall notify you after solving the pending pull request.

BTW, I am ditching my cygwin installation. From now on, I will test wslbridge2 in my msys2 installation instead of cygwin. cygwin has some weird issue with file permissions. If I reinstall Windows 10 the whole cygwin folder becomes inaccessible. But the wslbridge2 releases will have cygwin binaries compiled by Appveyor CI.

**Comment by PrincessRTFM** [comment]
> BTW, if anyone is using Windows 10 builds older than 20211, please test it also.

@Biswa96 `ver` in a windows command prompt returns `Microsoft Windows [Version 10.0.19041.508]` for me. I tried going to the CI page, opening the ["Image: Visual Studio 2019" job](https://ci.appveyor.com/project/Biswa96/wslbridge2/build/job/bps425a92u7uqclp), downloading the `wslbridge2.zip` file from the [artifacts tab](https://ci.appveyor.com/project/Biswa96/wslbridge2/build/job/bps425a92u7uqclp/artifacts), and replacing the `wslbridge2.exe` and `wslbridge-backend` files in my `AppData\Local\wsltty\bin` folder with the two from the zipfile. No luck. Quitting my shell normally and detaching my tmux session both give me the `/bin/wslbridge2: Exit 1` error line. Tried restoring the old `wslbridge-backend` file since that one wasn't mentioned, same thing. I'm using wsltty 3.3.0, the latest on the github releases page.

**Comment by Biswa96** [comment]
I have not build that commit in Appveyor. Please try the binaries from [wslbridge2 release page][1].

[1]: https://github.com/Biswa96/wslbridge2/releases

**Comment by PrincessRTFM** [comment]
Worked perfectly, thank you!

**Comment by Biswa96** [comment]
Replying to previous comment:

> Is GetVmId essential to shell invocation?
>>Though there are two possible way to circumvent the COM usage.

There is another way to circumvent the usage of Hyper-V sockets in WSL2. It _may be possible_ to use the IP address of utility virtual machine. But one has to add wslbridge2 in Windows Firewall which again requires admin permission.

**Comment by Biswa96** [comment]
This is fixed in wsltty version 3.4.1.
